DRAFT: bpf: use configurable large buffers for HTTP requests

mmat11 opened the pull request:
Context: https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/issues/396

rafaelroquetto commented on bpf/generictracer/protocol_http.h:
```diff
@@ -21,6 +23,11 @@
 
 volatile const u32 high_request_volume;
 
+enum {
+    k_http_large_buf_max_size = 1 << 14, // 16K
+    k_http_large_buf_max_size_mask = k_http_large_buf_max_size - 1,
+};
```
I think you don't need these. Instead, you can use `SCRATCH_MEM_NAMED(http_large_buffers, tcp_large_buffer_t)`

Then, in userspace, you effectively set the value size of this map to be that of the http buffer (if > 0) like you for `http_buffer_size`.

We do a similar thing here: https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/blob/f3dd462c1392aa433ffb014876230557b4ed77d1/pkg/components/netolly/ebpf/sock_tracer.go#L111

But instead of `.MaxEntries` you can use `.ValueSize`

mmat11 replied:
I could do this, but we don't save much per cpu, I wonder if it's worth adding this complication for a few kbytes

mmat11 replied:
Ok I am getting confused :D, the sizes specified there act as a "guard" for the maximum buffer size the event can assume regardless of config options. The event real size will always be dynamic depending on the payload, so if we get http events with a 20 bytes payload, we send 20 bytes + "event size" via the ringbuffer, not the max size (16k). I refactored everything large buffer related into `common/large_buffers.h` into a single constant (instead of per-protocol constants) and added a comment to make it clear. Let me know if this is good enough

rafaelroquetto commented on bpf/generictracer/protocol_http.h:
```diff
@@ -401,6 +408,52 @@ static __always_inline void handle_http_response(unsigned char *small_buf,
+    req->has_large_buffers = true;
+
+    bpf_ringbuf_output(
+        &events, large_buf, total_size & k_http_large_buf_max_size_mask, get_flags());
```
You may not need the mask here (in my tests I did not, but I did not test with older kernels).

If we can remove it, then we can do without those `k_http_large_buf...` constants altogether and rely on a single source of truth - way less error prone.

mmat11 replied:
I will test it without mask, I just put it out of habit :D

rafaelroquetto commented on pkg/components/ebpf/common/httpfltr_transform.go:
```diff
@@ -75,15 +76,31 @@ func ReadHTTPInfoIntoSpan(record *ringbuf.Record, filter ServiceFilter) (request
+		}
+		requestBuffer = b
+
+		// TODO: response?
```
Does that mean we are not using/consuming the buffers being set for the response? If that's the case, then we should remove that part from the ebpf code as well (i.e. not send any response buffers) until this is done - otherwise we are just wasting processing.

rafaelroquetto commented on pkg/components/ebpf/common/httpfltr_transform.go:
```diff
@@ -92,34 +109,34 @@ func HTTPInfoEventToSpan(event *BPFHTTPInfo) (request.Span, bool, error) {
 	}
-	result.URL = event.url()
-	result.Method = event.method()
+	result.URL = httpURLFromBuf(requestBuffer)
```
For each of these extraction methods, we are always allocating a new string no matter what e.g. `buf := string(req)` - which copies the underlying buffer, just to discard it sometimes when we do not find what we are looking for.

Instead we should strive to only allocate when we really need to return a non empty-value. This predates this PR but since we are touching these, it's a good opportunity to fix them.

mmat11 replied:
I'm refactoring this code in the next PR, can we address this there?

rafaelroquetto reviewed (commented):
Good stuff!

Just a bunch of minor comments. The more pressing of them IMHO relates to streamlining the buffer sizes - currently we have a lot of them, and it's not clear to the naked eyes / future maintainer how (and if) they relate.

My recommendation is to have the three sizes (HTTP, MySQL and Postgres) as the source of their respective truths.

It's then possible to adjust the respective map value sizes at load time to match those, and also to be a bit more lax on the accepted values: instead of hardcoding them, we can constrain them to the required alignment and adjust the final value - this is less error prone because everything is enforced as a rule, rather than explicit comparisons, analogous to https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/blob/f3dd462c1392aa433ffb014876230557b4ed77d1/pkg/obi/network_cfg.go#L97

mmat11 commented on bpf/generictracer/protocol_http.h:
```diff
@@ -401,6 +408,52 @@ static __always_inline void handle_http_response(unsigned char *small_buf,
+    large_buf->len = bytes_len;
+    if (large_buf->len >= http_buffer_size) {
+        large_buf->len = http_buffer_size;
+        bpf_dbg_printk("WARN: http_send_large_buffer: buffer is full, truncating data");
```
We don't -- we just truncate it and it's up to the handler to deal with it. I was thinking to introduce some new tag for this.
Imho we shouldn't drop it, there could be cases in troubleshooting in which it might be useful to see a span with truncated data rather than nothing

rafaelroquetto replied:
"Provide mechanism, not policy" they said! Very good point, thanks for clarifying.

rafaelroquetto commented on bpf/generictracer/iovec_len.h:
```diff
@@ -3,4 +3,4 @@
 #pragma once
 
-enum { k_iovec_max_len = 512 };
+enum { k_iovec_max_len = 8192 };
```
All good!

rafaelroquetto commented on bpf/common/large_buffers.h:
```diff
@@ -0,0 +1,24 @@
+    k_large_buf_max_size_mask = k_large_buf_max_size - 1,
+};
+
+SCRATCH_MEM_SIZED(http_large_buffers, k_large_buf_max_size);
```
Awesome!

rafaelroquetto commented on pkg/config/ebpf_tracer.go:
```diff
@@ -102,24 +102,27 @@ type EBPFTracer struct {
 	// TODO(matt): validate all the existing attributes
 
+	switch c.BufferSizes.HTTP {
+	case 0, 128, 256, 512, 1024, 2048, 4096, 8192:
```
I am suggesting to scrap this from Validate() altogether, and instead let the user specify any buffer size - we then adjust this value to the next power of 2 or next sane value:

```
// ensures http_buffer_size is a power of 2 of at least os.Pagesize()
m["http_buffer_size"] = effectiveRingBufferSize(p.cfg.EBPF.BufferSizes.HTTP)
```

(obviously we should likely rename `effectiveRingBufferSize()` to something like `effectiveBufferSize()`).

The generalisation comment was in reference to generalising `effectiveRingBufferSize()` for this purpose.

rafaelroquetto replied:
That's right and we should document that - perhaps even print it in the logs.

rafaelroquetto replied:
This got me wondering - does the buffer need to be a power of 2, or just aligned?

grcevski commented on pkg/components/ebpf/common/httpfltr_transform.go:
```diff
@@ -75,15 +76,29 @@ func ReadHTTPInfoIntoSpan(record *ringbuf.Record, filter ServiceFilter) (request
+	)
+
+	if event.HasLargeBuffers == 1 {
+		b, ok := extractTCPLargeBuffer(parseCtx, event.Tp.TraceId, event.Tp.SpanId, packetTypeRequest)
```
Actually, can we please change this to fallback to the event.Buf[:] if we fail to extract the buffer? No need to reject the span. I am thinking this is the reason the integration tests are failing.

mmat11 replied:
Yes I can do that, I am not sure the failing test is related though, the same test just failed in my other PR -> https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/actions/runs/18014863416/job/51257072591 😞

grcevski replied:
Yeah possibly not, it passed on the latest re-run, I'll take a look what's up. Maybe just make the safe fix anyway so that we get the most spans, better something than nothing.

mmat11 commented:
@rafaelroquetto @grcevski I removed the pow2 limitation so now the size can be anything from 0 to 8192. I also added fallback to the small buffer that was copied previously

mmat11 commented:
> Argh, I was about to hit merge and we have conflicts now. Sorry @mmat11, would it be possible to resolve them? I don't know which merge caused it.

Done, CI is green too ✅
